tools/build_file_generator/JavaClassAnalyzer.java

```diff
@@ -114,6 +233,53 @@ private static Import getImport(String importLine) {
     return result.build();
   }
 
+  private List<String> getUsedClassnamesInCode(String fileContent) {
+    final String multilineCommentRegex = "/\\*(?:.|[\\n\\r])*?\\*/";
+    final String stringValueRegex = "\"(.*?)\"";
+
+    List<String> javaCodeLines =
+        Arrays.stream(
+                fileContent
+                    .replaceAll(multilineCommentRegex, "")
+                    .replaceAll(stringValueRegex, "")
+                    .split(System.lineSeparator()))
+            .filter(line -> !line.trim().startsWith("//") && !line.trim().startsWith("import "))
+            .collect(Collectors.toList());
+
+    List<String> innerClasses =
+        getJavaClassnames(
+            javaCodeLines
+                .stream()
+                .filter(
+                    line ->
+                        line.contains(" class ")
+                            || line.contains(" interface ")
+                            || line.contains(" enum "))
+                .collect(Collectors.toList()));
+
+    return getJavaClassnames(javaCodeLines)
+        .stream()
+        .filter(classname -> !innerClasses.contains(classname))
+        .collect(Collectors.toList());
+  }
+
+  private List<String> getJavaClassnames(List<String> javaCodeLines) {
+    List<String> result = new ArrayList<>();
+    final String classnameRegex = "\\s([A-Z][a-z0-9|A-Z]+)+[\\s|.]";
+
+    for (String line : javaCodeLines) {
+      Pattern pattern = Pattern.compile(classnameRegex);
+      Matcher matcher = pattern.matcher(line);
+      while (matcher.find()) {
+        String classname = matcher.group().trim().replace(".", "");
+        if (!javaLangClasses.contains(classname) && !result.contains(classname)) {
```

Comment:
You can use a `TreeSet` (`SortedSet`):
https://docs.oracle.com/javase/8/docs/api/java/util/TreeSet.html

That way you avoid the `!result.contains(classname)` and it's clear to the reader that you only care about uniqueness.

Not sure if you need it sorted. You can also use `HashSet` and `Set` if not.

If you don't plan to change this list outside of this method, you can then return an `ImmutableList.copyOf(result)` and change the return type to `ImmutableList<String>`.

Reply:
Thank you! I changed it to `Set`.